### ISSUE TITLE
explain-pause-mode--enable-hooks fails when clicking echo area to open messages buffer (root-cause: (explain-pause-mode 1) more then once causes init hook code to run more then once)

### ISSUE THREAD

**Issue by wasamasa** [lastquestion/explain-pause-mode#78]
I updated to HEAD using Quelpa, restarted the Emacs daemon, clicked the echo area and was greeted by this:

```
Explain-pause-mode: please report this bug by creating a Github
issue at https://github.com/lastquestion/explain-pause-mode. Explain-pause-mode
is now _disabled_ so you can continue to hopefully use Emacs. Info follows:

explain-pause version: 0.1
emacs version: 26.3

wrap hook explain-pause-mode--enable-hooks for post-command-hook
current
#s(explain-pause-command-record root-emacs nil nil 6168 (24337 18204 551412 248000) nil nil nil nil 0)
should be equal
#s(explain-pause-command-record explain-pause-mode--enable-hooks nil #s(explain-pause-command-record post-command-hook nil #s(explain-pause-command-record root-emacs nil nil 6168 (24337 18204 551412 248000) nil nil nil nil 0) 0 nil nil nil nil nil 1) 3 (24337 18204 551418 997000) nil nil nil nil 2)

Backtrace:
  (explain-pause-report-measuring-bug "wrap hook explain-pause-mode--enable-hooks for post-command-hook" "current" #s(explain-pause-command-record root-emacs nil nil 6168 (24337 18204 551412 248000) nil nil nil nil 0) "should be equal" #s(explain-pause-command-record explain-pause-mode--enable-hooks nil #s(explain-pause-command-record post-command-hook nil #s(explain-pause-command-record root-emacs nil nil 6168 (24337 18204 551412 248000) nil nil nil nil 0) 0 nil nil nil nil nil 1) 3 (24337 18204 551418 997000) nil nil nil nil 2))
  (#[0 "\303\211\211\306\306H\307\310 \311H\"\211@\312_A@\\\313_\3148\313\245\\\262\\I\266\304\203\332\0\303\315 \210\211\306H
V\203F\0\211\211\316\317\320\321\322\310 \323\324 &I\266\211*\205\327\0\325\326\327\330H!\316H\331=?#\3321\224\0+\333H+\330H=\204\211\0\334*+\211\333H\314\\H\"\210+\211\333+\333HT,\246I\266\202^\0\334*\"0\266\203\202\331\0\211@\335=\203\240\0\336 \202\325\0\211@\337=\205\325\0+\330HT,\246+\333H=\203\275\0\336 \202\323\0+\211\330H\314\\I\266+\211\330I\262\262\266\202\266\202\210\303*\205v\325\340\327\330H!\327\341H\330H!\306H\316H\205\370\0\342%\33213+\333H+\330H=\204(\334*+\211\333H\314\\H\"\210+\211\333+\333HT,\246I\266\202\375\0\334*\"0\266\202\202x\211@\335=\203?\336 \202t\211@\337=\205t+\330HT,\246+\333H=\203\\\336 \202r+\211\330H\314\\I\266+\211\330I\262\262\266\202\262\210-\303=\204\215\343\325\344\300\301#\345-\346\303%\207\303\211\211\347\306H
VI\266\350\351\"\266\302\211\311\310 I\266\302\211-\207" [explain-pause-mode--enable-hooks post-command-hook #s(explain-pause-command-record root-emacs nil nil 6168 (24337 18204 551412 248000) nil nil nil nil 0) #s(explain-pause-command-record explain-pause-mode--enable-hooks nil #s(explain-pause-command-record post-command-hook nil #s(explain-pause-command-record root-emacs nil nil 6168 (24337 18204 551412 248000) nil nil nil nil 0) 0 nil nil nil nil nil 1) 3 (24337 18204 551418 997000) nil nil nil nil 2) nil explain-pause-slow-too-long-ms 4 time-subtract current-time 5 65536 1000 2 profiler-cpu-stop 9 profiler-make-profile :type cpu :timestamp :log profiler-cpu-log format "(\"profile-end\" \"%s\" %s)
" explain-pause--command-as-string 1 nil (file-error) 0 process-send-string file-missing explain-pause-log--missing-socket-disable file-error "(\"exit\" \"%s\" \"%s\" %s %s)
" 3 profile explain-pause-report-measuring-bug "wrap hook %s for %s" "current" "should be equal" 6 run-hook-with-args explain-pause-measured-command-hook explain-pause-log--send-process explain-pause-log--dgram-buffer explain-pause-log--dgram-buffer-size explain-pause--current-command-record] 11])
  (#[385 "
\204	\0\306\"\207\211\211\211\307\307H\310\311 \312H\"\211@\313_A@\\\314_\3158\314\245\\\262\\I\266\211\301\316\317\320\321\316\211\211\322H\316
\323HT&\266\204\211\300\316\317\320\321\316\211\211\322H\316
\323HT&\266\204\262
\205\324\325\326\327H!\326\327H!\326\330H\327H!\315H\307H\331H\332H	\322H
\323H&
\3331\323\0*\321H*\327H=\204\311\0\334
*\211\321H\315\\H\"\210*\211\321*\321HT+\246I\266\202\237\0\334
\"0\266\203\202\211@\335=\203\337\0\336 \202\211@\337=\205*\327HT+\246*\321H=\203\374\0\336 \202*\211\327H\315\\I\266*\211\327I\262\262\266\202\266\202\210\211\211,\205=\211\315H?\205=\211\322H?\205=\211\340\327H-.#\327H\262\262\211\203\352\211
\205\323\324\341\326\327H!\340\327H-\"#\3331\220*\321H*\327H=\204\206\334
*\211\321H\315\\H\"\210*\211\321*\321HT+\246I\266\202\\\334
\"0\266\202\202\325\211@\335=\203\234\336 \202\321\211@\337=\205\321*\327HT+\246*\321H=\203\271\336 \202\317*\211\327H\315\\I\266*\211\327I\262\262\266\202\262\210\211\211\332\342I\266\211\211\322\342I\266\343/!\266\211\312\311 I\266\344\321\345\346\347\300\301			%\350\"\351$\216\306\")\262\262\207" [explain-pause-mode--enable-hooks post-command-hook explain-pause-mode explain-pause--current-command-record explain-pause-root-command-loop explain-pause-log--send-process apply 4 time-subtract current-time 5 65536 1000 2 nil record explain-pause-command-record 0 8 10 format "(\"enter\" \"%s\" \"%s\" \"%s\" %s %s %s %s %s %d)
" explain-pause--command-as-string 1 3 6 7 (file-error) process-send-string file-missing explain-pause-log--missing-socket-disable file-error gethash "(\"profile-start\" \"%s\" %s)
" t profiler-cpu-start make-byte-code "\303\211\211\306\306H\307\310 \311H\"\211@\312_A@\\\313_\3148\313\245\\\262\\I\266\304\203\332\0\303\315 \210\211\306H
V\203F\0\211\211\316\317\320\321\322\310 \323\324 &I\266\211*\205\327\0\325\326\327\330H!\316H\331=?#\3321\224\0+\333H+\330H=\204\211\0\334*+\211\333H\314\\H\"\210+\211\333+\333HT,\246I\266\202^\0\334*\"0\266\203\202\331\0\211@\335=\203\240\0\336 \202\325\0\211@\337=\205\325\0+\330HT,\246+\333H=\203\275\0\336 \202\323\0+\211\330H\314\\I\266+\211\330I\262\262\266\202\266\202\210\303*\205v\325\340\327\330H!\327\341H\330H!\306H\316H\205\370\0\342%\33213+\333H+\330H=\204(\334*+\211\333H\314\\H\"\210+\211\333+\333HT,\246I\266\202\375\0\334*\"0\266\202\202x\211@\335=\203?\336 \202t\211@\337=\205t+\330HT,\246+\333H=\203\\\336 \202r+\211\330H\314\\I\266+\211\330I\262\262\266\202\262\210-\303=\204\215\343\325\344\300\301#\345-\346\303%\207\303\211\211\347\306H
VI\266\350\351\"\266\302\211\311\310 I\266\302\211-\207" vconcat vector [explain-pause-slow-too-long-ms 4 time-subtract current-time 5 65536 1000 2 profiler-cpu-stop 9 profiler-make-profile :type cpu :timestamp :log profiler-cpu-log format "(\"profile-end\" \"%s\" %s)
" explain-pause--command-as-string 1 nil (file-error) 0 process-send-string file-missing explain-pause-log--missing-socket-disable file-error "(\"exit\" \"%s\" \"%s\" %s %s)
" 3 profile explain-pause-report-measuring-bug "wrap hook %s for %s" "current" "should be equal" 6 run-hook-with-args explain-pause-measured-command-hook explain-pause-log--send-process explain-pause-log--dgram-buffer explain-pause-log--dgram-buffer-size explain-pause--current-command-record] 11 explain-pause-log--dgram-buffer explain-pause-log--dgram-buffer-size explain-pause-profile-enabled explain-pause-profile--profile-statistics explain-pause-profile--statistic-defaults explain-pause-profile-cpu-sampling-interval] 21 "

(fn ORIG-FUNC &rest ARGS)"] #[0 "\302\303\304\"\210\3051R\0\300\242\306V\203\0\307\310!\210\311\211\202P\0\312C\313\314\315\316\317\320!\321\"\322\323%\304\"\210\211\242\204G\0\324 \325=\203?\0\300\211\242T\240\210\326\303\304\312#\202N\0\211\242\205N\0\327 \2620\207\330\331\332#\207" [(1) explain-pause-mode remove-hook post-command-hook explain-pause-mode--enable-hooks (error) 5 message "Unable to install `explain-pause-mode'. please report a bug to github.com/lastquestion/explain-pause-mode" nil t mapbacktrace make-byte-code 1028 "\301=\206	\0\302=?\205\0\300\303\240\207" vconcat vector [explain-pause-mode--enable-hooks explain-pause-mode--try-enable-hooks nil] 6 "

(fn EVALED FUNC ARGS FLAGS)" recursion-depth 0 add-hook explain-pause-mode--install-hooks explain-pause-report-measuring-bug "Installation of explain-pause-mode failed" "error"] 8 "Install hooks for `explain-pause-mode' if it is being run at the top of the
emacs loop, e.g. not inside `call-interactively' or `sit-for' or any interleaved
timers, etc. Otherwise, try again."])
  (apply #[385 "
\204	\0\306\"\207\211\211\211\307\307H\310\311 \312H\"\211@\313_A@\\\314_\3158\314\245\\\262\\I\266\211\301\316\317\320\321\316\211\211\322H\316
\323HT&\266\204\211\300\316\317\320\321\316\211\211\322H\316
\323HT&\266\204\262
\205\324\325\326\327H!\326\327H!\326\330H\327H!\315H\307H\331H\332H	\322H
\323H&
\3331\323\0*\321H*\327H=\204\311\0\334
*\211\321H\315\\H\"\210*\211\321*\321HT+\246I\266\202\237\0\334
\"0\266\203\202\211@\335=\203\337\0\336 \202\211@\337=\205*\327HT+\246*\321H=\203\374\0\336 \202*\211\327H\315\\I\266*\211\327I\262\262\266\202\266\202\210\211\211,\205=\211\315H?\205=\211\322H?\205=\211\340\327H-.#\327H\262\262\211\203\352\211
\205\323\324\341\326\327H!\340\327H-\"#\3331\220*\321H*\327H=\204\206\334
*\211\321H\315\\H\"\210*\211\321*\321HT+\246I\266\202\\\334
\"0\266\202\202\325\211@\335=\203\234\336 \202\321\211@\337=\205\321*\327HT+\246*\321H=\203\271\336 \202\317*\211\327H\315\\I\266*\211\327I\262\262\266\202\262\210\211\211\332\342I\266\211\211\322\342I\266\343/!\266\211\312\311 I\266\344\321\345\346\347\300\301			%\350\"\351$\216\306\")\262\262\207" [explain-pause-mode--enable-hooks post-command-hook explain-pause-mode explain-pause--current-command-record explain-pause-root-command-loop explain-pause-log--send-process apply 4 time-subtract current-time 5 65536 1000 2 nil record explain-pause-command-record 0 8 10 format "(\"enter\" \"%s\" \"%s\" \"%s\" %s %s %s %s %s %d)
" explain-pause--command-as-string 1 3 6 7 (file-error) process-send-string file-missing explain-pause-log--missing-socket-disable file-error gethash "(\"profile-start\" \"%s\" %s)
" t profiler-cpu-start make-byte-code "\303\211\211\306\306H\307\310 \311H\"\211@\312_A@\\\313_\3148\313\245\\\262\\I\266\304\203\332\0\303\315 \210\211\306H
V\203F\0\211\211\316\317\320\321\322\310 \323\324 &I\266\211*\205\327\0\325\326\327\330H!\316H\331=?#\3321\224\0+\333H+\330H=\204\211\0\334*+\211\333H\314\\H\"\210+\211\333+\333HT,\246I\266\202^\0\334*\"0\266\203\202\331\0\211@\335=\203\240\0\336 \202\325\0\211@\337=\205\325\0+\330HT,\246+\333H=\203\275\0\336 \202\323\0+\211\330H\314\\I\266+\211\330I\262\262\266\202\266\202\210\303*\205v\325\340\327\330H!\327\341H\330H!\306H\316H\205\370\0\342%\33213+\333H+\330H=\204(\334*+\211\333H\314\\H\"\210+\211\333+\333HT,\246I\266\202\375\0\334*\"0\266\202\202x\211@\335=\203?\336 \202t\211@\337=\205t+\330HT,\246+\333H=\203\\\336 \202r+\211\330H\314\\I\266+\211\330I\262\262\266\202\262\210-\303=\204\215\343\325\344\300\301#\345-\346\303%\207\303\211\211\347\306H
VI\266\350\351\"\266\302\211\311\310 I\266\302\211-\207" vconcat vector [explain-pause-slow-too-long-ms 4 time-subtract current-time 5 65536 1000 2 profiler-cpu-stop 9 profiler-make-profile :type cpu :timestamp :log profiler-cpu-log format "(\"profile-end\" \"%s\" %s)
" explain-pause--command-as-string 1 nil (file-error) 0 process-send-string file-missing explain-pause-log--missing-socket-disable file-error "(\"exit\" \"%s\" \"%s\" %s %s)
" 3 profile explain-pause-report-measuring-bug "wrap hook %s for %s" "current" "should be equal" 6 run-hook-with-args explain-pause-measured-command-hook explain-pause-log--send-process explain-pause-log--dgram-buffer explain-pause-log--dgram-buffer-size explain-pause--current-command-record] 11 explain-pause-log--dgram-buffer explain-pause-log--dgram-buffer-size explain-pause-profile-enabled explain-pause-profile--profile-statistics explain-pause-profile--statistic-defaults explain-pause-profile-cpu-sampling-interval] 21 "

(fn ORIG-FUNC &rest ARGS)"] #[0 "\302\303\304\"\210\3051R\0\300\242\306V\203\0\307\310!\210\311\211\202P\0\312C\313\314\315\316\317\320!\321\"\322\323%\304\"\210\211\242\204G\0\324 \325=\203?\0\300\211\242T\240\210\326\303\304\312#\202N\0\211\242\205N\0\327 \2620\207\330\331\332#\207" [(1) explain-pause-mode remove-hook post-command-hook explain-pause-mode--enable-hooks (error) 5 message "Unable to install `explain-pause-mode'. please report a bug to github.com/lastquestion/explain-pause-mode" nil t mapbacktrace make-byte-code 1028 "\301=\206	\0\302=?\205\0\300\303\240\207" vconcat vector [explain-pause-mode--enable-hooks explain-pause-mode--try-enable-hooks nil] 6 "

(fn EVALED FUNC ARGS FLAGS)" recursion-depth 0 add-hook explain-pause-mode--install-hooks explain-pause-report-measuring-bug "Installation of explain-pause-mode failed" "error"] 8 "Install hooks for `explain-pause-mode' if it is being run at the top of the
emacs loop, e.g. not inside `call-interactively' or `sit-for' or any interleaved
timers, etc. Otherwise, try again."] nil)
  (explain-pause-mode--enable-hooks)
```

Another way of reproducing this: Eval code using M-: and after pressing RET a buffer with the above information pops up.

**Comment by lastquestion** [comment]
@wasamasa thank you for reporting this bug, :( 

HAH! I managed to reproduce this. I was thinking how it might be possible that this state could be reached.. until!

Can you check that your `init.el` or otherwise when you start with daemon, etc. you must call `(explain-pause-mode 1)` TWICE?

I can reproduce this when the mode is enabled and then forced enabled via `(explain-pause-mode 1)` AGAIN, with exactly the same backtrace.

I will fix it so that it doesn't attempt to install again if it's already installed, even if the user deliberately calls it with `1` argument, but can you check that in fact my guess is correct. :) If so, this should be fixed by only calling the mode once.

**Comment by wasamasa** [comment]
I'm using `after-init-hook` to enable it, so it should only be run once.

**Comment by lastquestion** [comment]
I got caught up with some stuff yesterday night, sorry for not getting a fix out. 

I agree that `after-init-hook` should only run once, but there's no other way I think this bug could fire. I checked `server.el`, and nothing jumps out ot me there. 

I will open a PR today for the crash when you call `(explain-pause-mode 1)` more then once. I think this should fix it. But if you're game for more debugging, try adding this very early - like very early, it doesn't need to have the library required - in your `init.el`

```elisp
(defadvice explain-pause-mode (before trace-it activate)
  (with-current-buffer "*Messages*"
    (mapbacktrace
     'backtrace--print-frame
     'explain-pause-mode)))
```

And that should log something like this in `*Messages*`
```
  explain-pause-mode()
  (lambda nil (explain-pause-mode))()
  run-hooks(after-init-hook delayed-warnings-hook)
  command-line()
  normal-top-level()
```
